src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java

```diff
@@ -1316,4 +1403,85 @@ private void setField(CpsFlowExecution result, String fieldName, Object value) {
      * this field is set to {@link CpsFlowExecution} that will own it.
      */
     static final ThreadLocal<CpsFlowExecution> PROGRAM_STATE_SERIALIZATION = new ThreadLocal<CpsFlowExecution>();
+
+    class TimingFlowNodeStorage extends FlowNodeStorage {
+        private final FlowNodeStorage delegate;
+        TimingFlowNodeStorage(FlowNodeStorage delegate) {
+            this.delegate = delegate;
+        }
+        @Override public FlowNode getNode(String string) throws IOException {
+            time(TimingKind.flow, false);
+            try {
+                return delegate.getNode(string);
+            } finally {
+                time(TimingKind.flow, true);
+            }
+        }
+        @Override public void storeNode(FlowNode fn) throws IOException {
+            time(TimingKind.flow, false);
+            try {
+                delegate.storeNode(fn);
+            } finally {
+                time(TimingKind.flow, true);
+            }
+        }
+        @Override public List<Action> loadActions(FlowNode node) throws IOException {
+            time(TimingKind.flow, false);
+            try {
+                return delegate.loadActions(node);
+            } finally {
+                time(TimingKind.flow, true);
+            }
+        }
+        @Override public void saveActions(FlowNode node, List<Action> actions) throws IOException {
+            time(TimingKind.flow, false);
+            try {
+                delegate.saveActions(node, actions);
+            } finally {
+                time(TimingKind.flow, true);
+            }
+        }
+    }
+
+    // If we wanted to expose via REST and/or floatingBox, could add a TransientActionFactory to show similar information.
+    @Extension(optional=true) public static class PipelineTimings extends Component {
```

Comment:
🔖  for me: I need to come back and give the components a deep review once the initial issues are resolved.